src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/PipelineExtensions.cs

```diff
@@ -68,7 +70,9 @@ public static ValueTask<ArraySegment<byte>> PeekAsync(this IPipeReader pipelineR
                 }
                 finally
                 {
-                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.Start);
+                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.IsEmpty
+                        ? result.Buffer.End
+                        : result.Buffer.Start);
                 }
 
                 readingTask = pipelineReader.ReadAsync();
```

Comment:
Wouldn't it be simpler to just put `pipelineReader.Advance(result.Buffer.End, result.Buffer.End)` right before this line? No need for any extra IsEmpty checks that way. If PeekAsync returns, ConsumedBytes should call Advance anyway, right?